gems/enumerize/2.5/_test/test.rbs

```diff
@@ -0,0 +1,3 @@
+class User
+  extend Enumerize::Predicates
```

Comment:
There is a diference between `test.rb` and `test.rbs`. In `test.rb`, `User` calls `extend` with `Enumerize`. But in `test.rbs`, `User` calls `extend` with `Enumerize::Predicates`.

I think this difference is confusing. When a model has `extend Enumerize`, the user should need to write only `extend Enumerize` to an rbs file. But the user needs to write `extend Enumerize::Predicates` actually with this patch.

Reply:
Please let me consult with you.

When we do `extend Enumerize` on `test.rbs`, we get the following error

```
test.rb:6:2: [error] Type `singleton(::User)` does not have method `enumerize`
│ Diagnostic ID: Ruby::NoMethod
│
└   enumerize :role, in: %i(user admin)
    ~~~~~~~~~

Detected 1 problem from 1 file
```

When we execute `extend Enumerize`, `Enumerize::Base.enumerize` is dynamically added to the calling class. I think that, this is why the above error occurs.

In this case, in `test.rbs`, I think I have no choice but to write it like `extend Enumerize::Base::ClassMethods`, is there any other better way?

Reply:
Thank you for the explanation.

Currently, RBS doesn't have any features to include/extend modules dynamically, so we cannot avoid differences between `.rbs` and `.rb`.

> In this case, in test.rbs, I think I have no choice but to write it like extend Enumerize::Base::ClassMethods, is there any other better way? As you say, it looks odd that test.rb and test.rbs are written differently.


I agree that we have no choice but to write the different `extend`. We cannot avoid this problem.

But I think we can reduce confusion among the end-users. Currently, the users are hard to know they need to write the different `extend`. The users will just write `extend Enumerize`, then they'll get the `Ruby::NoMethod` error. This is not a good experience.
So it needs something to help the users, such as documentation and/or a tool that automatically adds `extend`s.

How about adding a README file to `gems/enumerize/2.5/` to describe this problem, and writing a documentation comment to `Enumerize` module? It will be helpful to the users.

----


By the way, I'm not sure the current inheritance is the best way. How about adding `extend Enumerize` to the RBS file?

Currently, the rbs file only has `extend Enumerize::Base::ClassMethods`, but actually, the `User` class's inheritance is the following:

```ruby
require 'enumerize'

class User
  extend Enumerize
end

pp User.ancestors
# => [User, Enumerize::Base, Object, Kernel, BasicObject]
pp User.singleton_class.ancestors
# => [Enumerize::Base::ClassMethods::Hook,
#     #<Class:User>,
#     Enumerize::ModuleAttributes,
#     Enumerize::Predicates,
#     Enumerize::Base::ClassMethods,
#     Enumerize,
#     #<Class:Object>,
#     #<Class:BasicObject>,
#     Class,
#     Module,
#     Object,
#     PP::ObjectMixin,
#     Kernel,
#     BasicObject]
```

So technically the RBS file should be the following:

```rbs
class User
  include Enumerize::Base
  extend Enumerize
  extend Enumerize::ModuleAttributes
  extend Enumerize::Predicates
  extend Enumerize::Base::ClassMethods
  extend Enumerize::Base::ClassMethods::Hook
end
```


(Technically, it also should have `Enumerize::ActiveRecordSupport` and so on in some cases.)


I think the full `extend`s list is overkill. Because the main purpose of these modules is to define `enumerize` method, and the method is defined by `Enumerize::Base::ClassMethods`. So the current RBS file is enough for 99% of users I guess.

So most of the `extend`s are not necessary, but I think `extend Enumerize` is necessary. Because the Ruby file has `extend Enumerize` explicitly, so missing `extend Enumerize` in the RBS file looks odd. WDYT?

Reply:
Thank you for your very kind reply!

> How about adding a README file to gems/enumerize/2.5/ to describe this problem, and writing a documentation comment to Enumerize module? It will be helpful to the users.

Adding a README file would be a great way to do this. This way, as you say, it would reduce user confusion. Thanks for the suggestion!

---

> So most of the extends are not necessary, but I think extend Enumerize is necessary. Because the Ruby file has extend Enumerize explicitly, so missing extend Enumerize in the RBS file looks odd. WDYT?

I fully agree with you. I think `extend Enumerize` should be written in the RBS file as well. I think it is reasonable to adding `extend Enumerize` and the `Enumerize::Base::ClassMethods` that defines the `enumerize` method in the RBS file.

```
class User
  extend Enumerize
  extend Enumerize::Base::ClassMethods
end
```

I’ll fix `test.rbs` look like this and document its use in the README as well.
Thank you for your specific suggestions.

Reply:
Fixed `test.rbs` and added README.md.